### PR TITLE
Add Javascript clientState helper.  Introduce Mocha+Chai JS test framework for dashboard. 

### PR DESCRIPTION
**Pull request by philbogle**
Add Javascript accessor functions for progress and lines client state.  Update header.js.erb to use these helpers.
### Konacha (Mocha + Chai) Test framework

Adds the Konacha test frameworks.  In order include unit tests for the functions, the CL integrates the Konacha test framework (mocha + chai + Rails asset pipeline).

In brief, to run all tests interactively in your browser browser of choice:
    bundle exec rake konacha:serve
then hit "http://localhost:3500" in the browser

To run dashboard JS unit tests on the command line: 
     bundle exec rake konacha:run

(`konacha:run` requires that you install phantomjs  E.g. `brew install phantomjs` on OSX)

### CODE REVIEW

File: dashboard/app/assets/javascripts/client_state.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+dashboard.clientState = {};
+
+/**
+ * Number of days before client state cookie expires.
+ * @type {number}
+ * @private
+ */
+dashboard.clientState.EXPIRY_DAYS = 365;
+
+dashboard.clientState.reset = function() {
```
Comment (mehalshah) [inline dashboard/app/assets/javascripts/client_state.js:26]: Is there more to the client state than progress and lines? (videos seen, etc). Do we plan on resetting those? If not, then specify what's getting reset here.

File: dashboard/app/assets/javascripts/client_state.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+ * Returns the progress attained for the given level from the cookie.
+ * @param {number} level The id of the level
+ * @returns {number}
+ */
+dashboard.clientState.levelProgress = function(level) {
+  var progressMap = dashboard.clientState.allLevelsProgress();
+  return progressMap[String(level)] || 0;
+};
```
Comment (mehalshah) [inline dashboard/app/assets/javascripts/client_state.js:39]: You can one-line this. Any reason not to?

File: dashboard/app/assets/javascripts/client_state.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+ * Returns the number of lines completed from the cookie.
+ * @returns {number}
+ */
+dashboard.clientState.lines = function() {
+  var linesStr = $.cookie('lines');
+  return linesStr ? Number(linesStr) : 0;
+};
+
+/**
+ * Addeds the number of completed lines.
```
Comment (mehalshah) [inline dashboard/app/assets/javascripts/client_state.js]: Adds the param to the number of written lines (or something similar, the function description isn't exactly right)

File: dashboard/config/initializers/konacha.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# config/initializers/konacha.rb
+if defined?(Konacha)
```
Comment (mehalshah) [inline dashboard/config/initializers/konacha.rb:2]: Why wouldn't Konacha be defined?

**Comment by mehalshah** [comment]
Comments are minor / more questions. Overall, LGTM.

File: dashboard/app/assets/javascripts/client_state.js (hunk 1 above)
Comment (philbogle) [inline dashboard/app/assets/javascripts/client_state.js:26]: I added the module comment I had previously forgotten and made a note there about future things we'll add.

File: dashboard/app/assets/javascripts/client_state.js (hunk 2 above)
Comment (philbogle) [inline dashboard/app/assets/javascripts/client_state.js:39]: The parens end up being hard to read.  I think in this the case two easy-to-read lines are better than one concise but hard to read line

return (dashboard.clientState.allLevelsProgress())[String(level)] || 0;

File: dashboard/config/initializers/konacha.rb (hunk 4 above)
Comment (philbogle) [inline dashboard/config/initializers/konacha.rb:2]: It's not defined in all environments (e.g. production)

File: dashboard/app/assets/javascripts/client_state.js (hunk 3 above)
Comment (philbogle) [inline dashboard/app/assets/javascripts/client_state.js]: Fixed.

**Comment by philbogle** [comment]
Responded to Mehal's comments, thanks!  Added additional tests for malformed cookies.

**Pull request merged**